security/advancedtls/provider.go

```diff
@@ -0,0 +1,196 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package advancedtls
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"time"
+
+	"google.golang.org/grpc/credentials/tls/certprovider"
+	"google.golang.org/grpc/grpclog"
+)
+
+const (
+	defaultInterval = 1 * time.Hour
+)
+
+var logger = grpclog.Component("channelz")
+
+// IdentityPemFileProviderOptions contains fields to be filled out by users
+// for obtaining identity private key and certificates from PEM files.
+type IdentityPemFileProviderOptions struct {
+	CertFile string
+	KeyFile  string
+	Interval time.Duration
+}
+
+// IdentityPemFileProvider implements certprovider.Provider.
+// It provides the most up-to-date identity private key
+// and certificates based on the input PEM files.
+type IdentityPemFileProvider struct {
+	distributor *certprovider.Distributor
+	closeFunc   func()
+}
+
+// RootPemFileProviderOptions contains fields to be filled out by users
+// for obtaining root certificates from PEM files.
+type RootPemFileProviderOptions struct {
+	TrustFile string
+	Interval  time.Duration
+}
+
+// RootPemFileProvider implements certprovider.Provider.
+// It provides the most up-to-date root certificates based on the input PEM files.
+type RootPemFileProvider struct {
+	distributor *certprovider.Distributor
+	closeFunc   func()
+}
+
+// NewIdentityPemFileProvider uses IdentityPemFileProviderOptions to construct a IdentityPemFileProvider.
+func NewIdentityPemFileProvider(o *IdentityPemFileProviderOptions) (*IdentityPemFileProvider, error) {
+	if o.CertFile == "" {
+		return nil, fmt.Errorf("users need to specify certificate file path in NewIdentityPemFileProvider")
+	}
+	if o.KeyFile == "" {
+		return nil, fmt.Errorf("users need to specify key file path in NewIdentityPemFileProvider")
+	}
+	// If interval is not set by users explicitly, we will set it to default interval.
+	if o.Interval == 0*time.Nanosecond {
+		o.Interval = defaultInterval
+	}
+	quit := make(chan bool)
+	provider := &IdentityPemFileProvider{}
+	provider.distributor = certprovider.NewDistributor()
+	// If the initial files are empty, we will set an empty KeyMaterial.
+	certFileSize, _ := getFileSize(o.CertFile)
+	keyFileSize, _ := getFileSize(o.KeyFile)
+	if certFileSize == 0 || keyFileSize == 0 {
+		provider.distributor.Set(&certprovider.KeyMaterial{}, nil)
+	}
+	// A goroutine to pull file changes.
+	go func() {
+		for {
+			// Read identity certs from PEM files.
+			identityCert, err := tls.LoadX509KeyPair(o.CertFile, o.KeyFile)
+			// If the reading produces an error, we will skip the update for this round and log the error.
+			// Note that LoadX509KeyPair will return error if file is empty,
+			// so there is no separate check for empty file contents.
+			if err != nil {
+				logger.Warning("tls.LoadX509KeyPair(o.CertFile, o.KeyFile) failed: %v", err)
```

Comment:
nit: you can use:
logger.Warning("tls.LoadX509KeyPair(%v, %v) failed: %v", o.CertFile, o.KeyFile, err)

Reply:
Sure, will fix!